java-squid/src/main/java/org/sonar/java/collections/AVLTree.java

```diff
@@ -0,0 +1,388 @@
+      AVLTree lrl = lr.left();
+      AVLTree lrr = lr.right();
+      return createNode(createNode(ll, l, lrl), lr, createNode(lrr, key, value, r));
+    }
+    if (r.height() > l.height() + 2) {
+      assert !r.isEmpty();
+      AVLTree rl = r.left();
+      AVLTree rr = r.right();
+      if (rr.height() >= rl.height()) {
+        return createNode(createNode(l, key, value, rl), r, rr);
+      }
+      assert !rl.isEmpty();
+      AVLTree rll = rl.left();
+      AVLTree rlr = rl.right();
+      return createNode(createNode(l, key, value, rll), rl, createNode(rlr, r, rr));
+    }
+    return createNode(l, key, value, r);
+  }
+
+  private static AVLTree createNode(AVLTree newLeft, AVLTree oldTree, AVLTree newRight) {
+    return createNode(newLeft, oldTree.key(), oldTree.value(), newRight);
+  }
+
+  private static AVLTree createNode(AVLTree l, Object key, Object value, AVLTree r) {
+    return new Node(l, r, key, value, incrementHeight(l, r));
+  }
+
+  private static int incrementHeight(AVLTree l, AVLTree r) {
+    return (l.height() > r.height() ? l.height() : r.height()) + 1;
+  }
+
+  private static class Node extends AVLTree {
+    private final AVLTree left;
+    private final AVLTree right;
+    private final int height;
+
+    private final Object key;
+    private final Object value;
+    private int hashCode;
+
+    public Node(AVLTree left, AVLTree right, Object key, Object value, int height) {
+      this.left = left;
+      this.right = right;
+      this.key = key;
+      this.value = value;
+      this.height = height;
+    }
+
+    @Override
+    protected AVLTree left() {
+      return left;
+    }
+
+    @Override
+    protected AVLTree right() {
+      return right;
+    }
+
+    @Override
+    protected Object key() {
+      return key;
+    }
+
+    @Override
+    protected Object value() {
+      return value;
+    }
+
+    @Override
+    protected int height() {
+      return height;
+    }
+
+    @Override
+    public boolean isEmpty() {
+      return false;
+    }
+
+    @Override
+    public int hashCode() {
+      if (hashCode == 0) {
+        int result = key.hashCode();
+        result = result * 31 + left.hashCode();
+        result = result * 31 + right.hashCode();
+        result = result * 31 + value.hashCode();
+        hashCode = result;
+      }
+      return hashCode;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (this == obj) {
+        return true;
+      }
+      if (obj instanceof Node) {
+        Node other = (Node) obj;
+        return Objects.equals(this.key, other.key)
+          && Objects.equals(this.value, other.value)
+          && Objects.equals(this.left, other.left)
+          && Objects.equals(this.right, other.right);
+      }
+      return false;
+    }
+
+    @Override
+    public String toString() {
+      return " " + key + "->" + value + left.toString() + right.toString();
+    }
+  }
+
+  private static final AVLTree EMPTY = new AVLTree() {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce this anonymous class number of lines from 46 to at most 20, or make it a named class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1188)